src/main/scala/cognite/spark/v1/AssetsHierarchyBuilder.scala

```diff
@@ -0,0 +1,221 @@
+    sourceRoot match {
+      case None =>
+        client.assets.create(
+          Seq(
+            AssetCreate(
+              root.name,
+              None,
+              root.description,
+              root.source,
+              Some(root.externalId),
+              root.metadata,
+              None
+            ))) *> IO.unit
+      case Some(asset) =>
+        val isEqual =
+          root.description == asset.description &&
+            root.metadata == asset.metadata &&
+            root.name == asset.name &&
+            root.source == asset.source
+        if (isEqual) {
+          IO.unit
+        } else {
+          client.assets.updateByExternalId(
+            Seq(
+              root.externalId -> AssetsIngestSchema
+                .toAssetUpdate(root)
+                .copy(parentExternalId = None)).toMap) *> IO.unit
+        }
+    }
+
+  def deleteInsertUpdate(
+      root: AssetsIngestSchema,
+      leaves: Set[AssetsIngestSchema],
+      cdfTree: Seq[Asset]): (Seq[Asset], Set[AssetsIngestSchema], Set[AssetsIngestSchema]) = {
+    // Insert assets that are not present in CDF
+    val toInsert = leaves.filter(a => !cdfTree.map(_.externalId.get).contains(a.externalId))
+
+    // Keep assets that are in both source and cdf, delete those only in CDF
+    val (toKeep, toDelete) =
+      cdfTree.partition(a => (Seq(root) ++ leaves).map(_.externalId).contains(a.externalId.get))
+
+    // Filter out assets that would not be changed in an update
+    val toUpdate = leaves.filter(a => needsUpdate(a, toKeep))
+
+    (toDelete, toInsert, toUpdate)
+  }
+
+  def needsUpdate(leaf: AssetsIngestSchema, cdfTree: Seq[Asset]): Boolean = {
+    val cdfAsset = cdfTree.find(a => a.externalId.contains(leaf.externalId))
+    cdfAsset match {
+      case Some(asset) =>
+        cdfTree.find(a => asset.parentId.contains(a.id)) match {
+          case Some(matchingAsset) =>
+            leaf.description != asset.description ||
+              leaf.metadata != asset.metadata ||
+              leaf.name != asset.name ||
+              leaf.source != asset.source ||
+              !matchingAsset.externalId.contains(leaf.parentExternalId)
+          case None =>
+            false
+        }
+      case None => false
+    }
+  }
+
+  def validateAndOrderInput(
+      tree: Array[AssetsIngestSchema]): (AssetsIngestSchema, Set[AssetsIngestSchema]) = {
+    val (root, leaves) = splitIntoRootAndLeaves(tree)
+
+    (root, isTree(root, leaves))
+  }
+
+  def splitIntoRootAndLeaves(
+      tree: Array[AssetsIngestSchema]): (AssetsIngestSchema, Array[AssetsIngestSchema]) = {
+    val (maybeRoots, leaves) = tree.partition(_.parentExternalId == "")
+
+    val root = maybeRoots.toList match {
+      case singleRoot :: Nil => singleRoot
+      case Nil => throw NoRootException()
+      case _ => throw MultipleRootsException()
+    }
+    (root, leaves)
+  }
+
+  def isTree(root: AssetsIngestSchema, leaves: Array[AssetsIngestSchema]): Set[AssetsIngestSchema] = {
```

Comment:
This should be named something else considering it returns a set of nodes. "is" should be for boolean return values.